### PR TITLE
complex number support

### PR DESCRIPTION
**Pull request by GiggleLiu**
Related Issues

#157 
#324 
#364 

This is a quick fix for some complex valued functions. Any thoughts?

NOTE: I copied some functions from Base, because they fixed the type to foating point numbers.

### CODE REVIEW

File: src/complex.jl [hunk 1]
```diff
@@ -0,0 +1,172 @@
+Base.prevfloat(x::Dual{T,V}) where {T,V<:AbstractFloat} = prevfloat(x.value)
+Base.nextfloat(x::Dual{T,V}) where {T,V<:AbstractFloat} = nextfloat(x.value)
+
+function Base.log(z::Complex{T}) where {A, FT<:AbstractFloat, T<:Dual{A,FT}}
+    T1::T  = 1.25
```
Comment (YingboMa) [inline src/complex.jl:5]:
```suggestion
    T1  = convert(T, 5//4)
```

File: src/complex.jl [hunk 2]
```diff
@@ -0,0 +1,172 @@
+Base.prevfloat(x::Dual{T,V}) where {T,V<:AbstractFloat} = prevfloat(x.value)
+Base.nextfloat(x::Dual{T,V}) where {T,V<:AbstractFloat} = nextfloat(x.value)
+
+function Base.log(z::Complex{T}) where {A, FT<:AbstractFloat, T<:Dual{A,FT}}
```
Comment (YingboMa) [inline src/complex.jl:4]:
```suggestion
function Base.log(z::Complex{T}) where {A, T<:Dual{A,<:AbstractFloat}}
```

File: src/complex.jl [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    T1::T  = 1.25
+    T2::T  = 3
```
Comment (YingboMa) [inline src/complex.jl:6]:
```suggestion
    T2  = convert(T, 3)
```

File: src/complex.jl [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    ln2::T = log(convert(T,2))  #0.6931471805599453
```
Comment (YingboMa) [inline src/complex.jl:7]: I don't think you need a type annotation here.

File: src/complex.jl (hunk 1 above)
Comment (YingboMa) [reply src/complex.jl:5]: I am not sure about this, because the compiler cannot optimize the rational number away.

File: test/ComplexTest.jl [hunk 5]
```diff
@@ -0,0 +1,53 @@
+using ForwardDiff: Dual
```
Comment (YingboMa) [inline test/ComplexTest.jl:1]: ForwardDiff wraps all test files into a module.

File: src/complex.jl [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        θ, β = ay, ax
+    end
+    if k==0 && (0.5 < β*β) && (β <= T1 || ρ < T2)
+        ρρ = log1p((β-1)*(β+1)+θ*θ)/2
+    else
+        ρρ = log(ρ)/2 + k*ln2
+    end
+    Complex(ρρ, angle(z))
+end
+function Base.tanh(z::Complex{T}) where {A, FT<:AbstractFloat, T<:Dual{A,FT}}
```
Comment (YingboMa) [inline src/complex.jl:24]:
```suggestion
function Base.tanh(z::Complex{T}) where {A, T<:Dual{A,<:AbstractFloat}}
```

File: src/complex.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/complex.jl:5]: This function is copied from Base. One of the goal of this PR is to stay consistent with the implementation in Base.

https://github.com/JuliaLang/julia/blob/575701248adc951038b9727e646230737e5b3c91/base/complex.jl#L564

File: test/ComplexTest.jl (hunk 5 above)
Comment (GiggleLiu) [reply test/ComplexTest.jl:1]: Yes, interesting design. May I know the reason to put tests in a module?

File: src/complex.jl (hunk 3 above)
Comment (GiggleLiu) [reply src/complex.jl:6]: same as above.

File: test/ComplexTest.jl (hunk 5 above)
Comment (GiggleLiu) [reply test/ComplexTest.jl:2]: I have updated the tests. Good now?

Comment (YingboMa) [reply test/ComplexTest.jl:2]: It won't introduce variables to `Main`, in this way. So tests don't influence each other. This is the motivation for https://github.com/YingboMa/SafeTestsets.jl, too.

**Comment by GiggleLiu** [comment]
@YingboMa 
Although we get only 3 votes in #456 , but I need to differentiate complex number in one of my recent projects.
Do you think we should take the chance to polish and get this PR merged?

File: src/complex.jl [hunk 7]
```diff
@@ -0,0 +1,172 @@
+Base.prevfloat(x::Dual{T,V}) where {T,V<:AbstractFloat} = prevfloat(x.value)
```
Comment (ChrisRackauckas) [inline src/complex.jl:1]: see https://github.com/JuliaDiff/ForwardDiff.jl/pull/453

Comment (GiggleLiu) [reply src/complex.jl:1]: Thanks. Is that PR going to be merged soon? If so, feel free to merge that PR first, I can do the compatibility check later.

**Comment by ChrisRackauckas** [comment]
Is this not all covered by https://github.com/JuliaLang/julia/pull/36030 ?

File: src/complex.jl (hunk 7 above)
Comment (ChrisRackauckas) [reply src/complex.jl:1]: done

**Comment by ChrisRackauckas** [comment]
Looking at the two PRs, it seems like @YingboMa upstreamed it all to Base where it should go (correct me if I'm wrong and anything was left out). So complex AD is thus fixed on v1.6. However, the tests are still definitely useful from here. So I'd suggest the following:

- Move these overloads to an internal module. Document in the README `using ForwardDiff.Complex` for v1.4 as a compatibility workaround.
- Use this to in the tests

and everything is good to go. The other alternative would be to move these behind an `@static if` definition, but I would be weary of having such central dispatches modified by such a core package outside of the system image. If you want invalidations, this is how you get invalidations.

**Comment by ChrisRackauckas** [comment]
Handled by https://github.com/JuliaLang/julia/pull/36030

**Comment by GiggleLiu** [comment]
Thanks! @ChrisRackauckas

**Comment by DEKOHG** [comment]
Hello,
I just installed the newest version of DifferentialEquations.jl . However, I still got an error with stiff solvers (eg. Rodas4P) for complex numbers calculation. I have tried on both julia 1.8.2 and julia 1.7.2 and my DifferentialEquations is v7.6.0

ERROR: ArgumentError: Cannot create a dual over scalar type ComplexF64. If the type behaves as a scalar, define ForwardDiff.can_dual(::Type{ComplexF64}) = true.

**Comment by petar-andrejic** [comment]
> Handled by [JuliaLang/julia#36030](https://github.com/JuliaLang/julia/pull/36030)

Not entirely handled, complex valued functions of multiple real arguments don't work, e.g.

```julia
julia> ForwardDiff.gradient(x->exp(im * (x[1] + x[2])), [1.0,2.0])

ERROR: DimensionMismatch: gradient(f, x) expects that f(x) is a real number. Perhaps you meant jacobian(f, x)?
```

**Comment by davidschlegel** [comment]
> Hello, I just installed the newest version of DifferentialEquations.jl . However, I still got an error with stiff solvers (eg. Rodas4P) for complex numbers calculation. I have tried on both julia 1.8.2 and julia 1.7.2 and my DifferentialEquations is v7.6.0
> 
> ERROR: ArgumentError: Cannot create a dual over scalar type ComplexF64. If the type behaves as a scalar, define ForwardDiff.can_dual(::Type{ComplexF64}) = true.

I also still get the same error. Is there a way to manually change the way forward differentiation is handled?

**Comment by ChrisRackauckas** [comment]
For DiffEq? You just set `autodiff=false` (`Rodas5P(autodiff=false)`) or pass a `jac` function into `ODEFunction` if you want full control of it.

**Comment by davidschlegel** [comment]
Sorry, it was not clear from my question. Of course, setting `autodiff=false` works but is not efficient.
Can I autodiff a complex ODE in DiffEq? Currently it does not work out-of-the-box, but perhaps one can manually pass a Jacobian that is calculated using ForwardDiff...

**Comment by ChrisRackauckas** [comment]
It uses `ForwardDiff` by default, and you can pass any Jacobian function you want via `jac` in the `ODEFunction`.

**Comment by davidschlegel** [comment]
Exactly, but right now, this throws an error, when trying to autodiff a complex ODE.

**Comment by ChrisRackauckas** [comment]
So then I don't get the question?

**Comment by davidschlegel** [comment]
Right now, afaik, ForwardDiff does not handle automatic differentiation for complex functions out-of-the-box. However, DiffEq relies on ForwardDiff when computing jacobians and more. Is there any solution to still use automatic differentiation (eg using a different package) for DiffEq?

**Comment by ChrisRackauckas** [comment]
Yes, just use `jac` in the `ODEFunction`.

**Pull request closed without merging**